src/backend/webmidi/mod.rs

```diff
@@ -0,0 +1,290 @@
+}
+
+pub struct MidiInput {
+    ignore_flags: Ignore
+}
+
+impl MidiInput {
+    pub fn new(_client_name: &str) -> Result<Self, InitError> {
+        Ok(MidiInput { ignore_flags: Ignore::None })
+    }
+
+    pub fn ignore(&mut self, flags: Ignore) {
+        self.ignore_flags = flags;
+    }
+
+    pub fn port_count(&self) -> usize {
+        STATIC.with(|s| {
+            let mut s = s.borrow_mut();
+            s.refresh_inputs();
+            s.input_set.len()
+        })
+    }
+
+    pub fn port_name(&self, port_number: usize) -> Result<String, PortInfoError> {
+        STATIC.with(|s| {
+            let s = s.borrow_mut();
+            if port_number >= s.input_set.len() { return Err(PortInfoError::PortNumberOutOfRange); }
+            let input = &s.input_set.list()[port_number];
+            Ok(input.name().unwrap_or_else(|| input.id()))
+        })
+    }
+
+    pub fn connect<F, T: Send + 'static>(
+        self, port_number: usize, _port_name: &str, mut callback: F, data: T
+    ) -> Result<MidiInputConnection<T>, ConnectError<MidiInput>>
+        where F: FnMut(u64, &[u8], &mut T) + Send + 'static
+    {
+        STATIC.with(|s| {
+            let s = s.borrow_mut();
+            if port_number >= s.input_set.len() { return Err(ConnectError::new(ConnectErrorKind::PortNumberOutOfRange, self)); }
+            let input = s.input_set.list()[port_number].clone();
+            let _ = input.open(); // NOTE: asyncronous!
+
+            let ignore_flags = self.ignore_flags;
+            let user_data = Arc::new(Mutex::new(Some(data)));
+
+            let closure = {
+                let user_data = user_data.clone();
+
+                let closure = Closure::wrap(Box::new(move |event: MidiMessageEvent| {
+                    let time = (event.time_stamp() * 1000.0) as u64; // ms -> us
+                    let buffer = event.data().unwrap();
+
+                    let status = buffer[0];
+                    if !(status == 0xF0 && ignore_flags.contains(Ignore::Sysex) ||
+                         status == 0xF1 && ignore_flags.contains(Ignore::Time) ||
+                         status == 0xF8 && ignore_flags.contains(Ignore::Time) ||
+                         status == 0xFE && ignore_flags.contains(Ignore::ActiveSense))
+                    {
+                        callback(time, &buffer[..], user_data.lock().unwrap().as_mut().unwrap());
+                    }
+                }) as Box<dyn FnMut(MidiMessageEvent)>);
+
+                input.set_onmidimessage(Some(closure.as_ref().unchecked_ref()));
+
+                closure
+            };
+
+            Ok(MidiInputConnection { ignore_flags, input, user_data, closure })
+        })
+    }
+}
+
+pub struct MidiInputConnection<T> {
+    ignore_flags:   Ignore,
+    input:          web_sys::MidiInput,
+    user_data:      Arc<Mutex<Option<T>>>,
+    #[allow(dead_code)] // Must be kept alive until we decide to unregister from input
+    closure:        Closure<dyn FnMut(MidiMessageEvent)>,
```

Comment:
If that is how you have to do things for event handlers in `js_sys`, that seems a bit unsafe to me ... but probably it's the only way?

Reply:
I just found https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/closure/struct.Closure.html which explain this a bit.

Reply:
Yeah, I don't have a great alternative here.  Dropping the closure will cause the JS function to start throwing exceptions/panicing when called.  Fortunately, this starts happening immediately, so it's easy to test/catch if mistakes are made.